### PR TITLE
Ben Kaduk's nits on HTTP/3

### LINKED ISSUE #4760: HTTP/[23] Parallels in IANA (aka Ben's HTTP/3 Comment 7)
@kaduk notes that the text for avoiding conflicting registrations with HTTP/2 doesn't actually empower the expert to refuse a registration if the registrant *really really* wants to stomp a value already assigned in the other protocol.  Add a sentence that expert reviewers MAY refuse a registration on those grounds.

### LINKED ISSUE #4776: Ben Kaduk's HTTP/3 Comment 1
@kaduk said:

> Thanks to Hilarie Orman for the secdir review and to the team for
> responding to it!  It looks like one of the secdir review comments
> didn't get its own github issue, so perhaps we can discuss it now:
>
> > I would like to see the Security Considerations spell out exactly
> > what security features HTTP expects from QUIC.
>
> Perhaps we could add a line to Section 1.2 where we discuss how QUIC
> provides comparable confidentiality and integrity to TLS+TCP, along the
> lines of "HTTP/3 relies on QUIC to provide that confidentiality and
> integrity protection of data, as well as to provide peer authentication
> and reliable, in-order, per-stream delivery."

### LINKED ISSUE #4777: Ben Kaduk's HTTP/3 Comment 2
@kaduk said

> With all HTTP/3 frames being sent on QUIC streams, including
> connection-wide control frames on the control stream, we are in the
> situation where QUIC flow control can prevent the conveyance of HTTP
> control messages.  This is in some sense covered in Appendix A.2.3,
> since "all HTTP/3 frame types defined in this document are sent on
> streams", but the corresponding PR discussion seems to have mostly
> focused on how this applies to request/response headers, and while the
> discussion in #204 does confirm that it is intentional for flow control
> to apply, that doesn't really cover whether it is worth noting as a
> potentially surprising difference from HTTP/2.  While the HTTP frames
> currently defined for the control stream might not be too catastrophic
> to be blocked from sending (assuming that SETTINGS will typically get
> through in the initial window), I do still wonder if giving specific
> guidance on flow control for the HTTP control stream could be useful.
> (E.g., calling out that an endpoint might terminate the connection if
> blocked by flow control on the control stream for an extended duration,
> or even requiring preferential treatment for window updates.)

### LINKED ISSUE #4780: Ben Kaduk's HTTP/3 Comment 5
@kaduk said

> Section 10.5.2
>
> There is some good discussion in [HTTP-SEMANTICS] about the risk of
> CONNECT being used as an arbitrary tunnel, that is probably worth
> referencing from here.
>
> >   The CONNECT method can be used to create disproportionate load on a
> >   proxy, since stream creation is relatively inexpensive when compared
> >   to the creation and maintenance of a TCP connection.  A proxy might
> >   also maintain some resources for a TCP connection beyond the closing
> >   of the stream that carries the CONNECT request, since the outgoing
> >   TCP connection remains in the TIME_WAIT state.  Therefore, a proxy
> >   cannot rely on QUIC stream limits alone to control the resources
> >   consumed by CONNECT requests.
>
> I'm not sure how well this last sentence translates from HTTP/2 to
> HTTP/3 -- in HTTP/2 the limit is on the number of concurrent streams,
> but QUIC gives a hard cap on the absolute stream number, so IIUC an
> endpoint could refuse to allow new streams until TCP state had quiesced.
> (Of course, this would also prevent any new HTTP requests from being
> sent...)

### LINKED ISSUE #4781: Ben Kaduk's HTTP/3 Comment 6
@kaduk said:

> Section 10.9
>
> I think it might be worth another sentence to reiterate that when
> [HTTP-REPLAY] talks about "the TLS layer", this applies to the TLS
> handshake layer as used as part of QUIC v1, and when it talks about
> "early data" that means QUIC 0-RTT.  FWIW, I made a quick pass through
> RFC 8470 to look for references to TLS, and didn't find any references
> to TLS that would not also apply to QUIC v1's usage of TLS.  (It does
> have things like "early data received on that TLS connection" that do
> not strictly speaking apply to HTTP/3, but I don't see that one as
> particularly problematic.)

### LINKED ISSUE #4782: Ben Kaduk's HTTP/3 Comment 8
@kaduk said

> Section 11.2.2
>
> I note that RFC 7540 marked setting code 0x0 reserved for h2; should we
> mirror that?

### LINKED ISSUE #4783: Ben Kaduk's HTTP/3 Comment 9
@kaduk said

> Section 12.1
>
> [ALTSVC] doesn't exactly seem to be referenced in a normative manner
> anymore.  Back when it was the only way to do HTTP/3 it made sense, but
> the current description seems mostly to be saying that "other services
> could use Alt-Svc to initiate HTTP/3" which is arguably not really
> something you have to know in order to implement and use this spec.

### LINKED ISSUE #4785: Ben Kaduk's HTTP/3 Comment 11
@kaduk said

> Appendix A.3
>
> (nit) I note that A.2.5 gave the codepoints for the HTTP/2 frame types being
> discussed; should do the analogous thing for HTTP/2 settings types?

### PR DESCRIPTION
**Pull request by MikeBishop**
Collection of editorial changes from @kaduk's review.

Fixes #4760.
Fixes #4776.
Fixes #4777.
Fixes #4780.
Fixes #4781.
Fixes #4782.
Fixes #4783 if no one objects.
Fixes #4785.

### CODE REVIEW

**Review by kaduk — APPROVED** [review]
Thanks; these look good to me.

File: draft-ietf-quic-http.md [hunk 1]
```diff
@@ -1908,13 +1920,13 @@ A client can discard responses that it cannot process.
 
 ### CONNECT Issues
 
-The CONNECT method can be used to create disproportionate load on a proxy,
-since stream creation is relatively inexpensive when compared to the creation
-and maintenance of a TCP connection.  A proxy might also maintain some resources
-for a TCP connection beyond the closing of the stream that carries the CONNECT
+The CONNECT method can be used to create disproportionate load on a proxy, since
```
Comment (martinthomson) [inline draft-ietf-quic-http.md:1923]:
```suggestion
The CONNECT method can be used to create disproportionate load on a proxy, as
```

File: draft-ietf-quic-http.md [hunk 2]
```diff
@@ -1908,13 +1920,13 @@ A client can discard responses that it cannot process.
...
+stream creation is relatively inexpensive when compared to the creation and
+maintenance of a TCP connection.  A proxy might also maintain some resources for
+a TCP connection beyond the closing of the stream that carries the CONNECT
 request, since the outgoing TCP connection remains in the TIME_WAIT state.
-Therefore, a proxy cannot rely on QUIC stream limits alone to control the
-resources consumed by CONNECT requests.
+Therefore, a proxy might delay increasing the QUIC stream limits to account for
+the resources consumed by CONNECT requests.
```
Comment (martinthomson) [inline draft-ietf-quic-http.md:1929]: The original text existed to note that accounting for streams as requests is insufficient in the case that a CONNECT request is used.  Delay does no good.  What we might instead say is that any proxy that supports CONNECT might choose to limit the number of concurrent CONNECT streams it accepts (and account for the higher cost of those streams when considering overall resource allocation).

File: draft-ietf-quic-http.md [hunk 3]
```diff
@@ -2267,6 +2286,10 @@ of time.  HTTP/3 servers might choose to permit a larger number of concurrent
 client-initiated bidirectional streams to achieve equivalent concurrency to
 HTTP/2, depending on the expected usage patterns.
 
+In HTTP/2, only request and response bodies (DATA frames) are subject to flow
```
Comment (martinthomson) [inline draft-ietf-quic-http.md:2289]:
```suggestion
In HTTP/2, only request and response bodies (the frame payload of DATA frames) are subject to flow
```

**Review by martinthomson — COMMENTED** [review]
One change here looks a little off.

File: draft-ietf-quic-http.md (hunk 2 above)
Comment (kaduk) [reply draft-ietf-quic-http.md:1929]: It looks like the original text did two things: note that CONNECT streams are more expensive than request streams (as you note here), but also that resources used for CONNECT streams might linger even after the HTTP stream is closed (the point I remarked upon where the properties of HTTP/3 and HTTP/2 diverge qualitatively).

Comment (MikeBishop) [reply draft-ietf-quic-http.md:1929]: Hm, that's a different reading that I'd initially taken.  I'll try to rework the text a bit more.

**Pull request merged**